Tools/build/generate_sbom.py

```diff
@@ -0,0 +1,179 @@
+    return sorted([line.split()[-1] for line in git_check_ignore_lines if line.startswith("::")])
+
+
+def main() -> None:
+    root_dir = pathlib.Path(__file__).parent.parent.parent
+    sbom_path = root_dir / "Misc/sbom.spdx.json"
+    sbom_data = json.loads(sbom_path.read_bytes())
+
+    # Make a bunch of assertions about the SBOM data to ensure it's consistent.
+    assert {package["name"] for package in sbom_data["packages"]} == set(PACKAGE_TO_FILES)
+    for package in sbom_data["packages"]:
+
+        # Properties and ID must be properly formed.
+        assert set(package.keys()) == REQUIRED_PROPERTIES_PACKAGE
+        assert package["SPDXID"] == spdx_id(f"SPDXRef-PACKAGE-{package['name']}")
+
+        # Version must be in the download and external references.
+        version = package["versionInfo"]
+        assert version in package["downloadLocation"]
+        assert all(version in ref["referenceLocator"] for ref in package["externalRefs"])
+
+        # License must be on the approved list for SPDX.
+        assert package["licenseConcluded"] in ALLOWED_LICENSE_EXPRESSIONS, package["licenseConcluded"]
+
+    # Regenerate file information from current data.
+    sbom_files = []
+    sbom_relationships = []
+
+    # We call 'sorted()' here a lot to avoid filesystem scan order issues.
+    for name, files in sorted(PACKAGE_TO_FILES.items()):
+        package_spdx_id = spdx_id(f"SPDXRef-PACKAGE-{name}")
+        exclude = files.exclude or ()
+        for include in sorted(files.include):
+
+            # Find all the paths and then filter them through .gitignore.
+            paths = glob.glob(include, root_dir=root_dir, recursive=True)
+            paths = filter_gitignored_paths(paths)
+            assert paths, include  # Make sure that every value returns something!
+
+            for path in paths:
+                # Skip directories and excluded files
+                if not (root_dir / path).is_file() or path in exclude:
+                    continue
+
+                # SPDX requires SHA1 to be used for files, but we provide SHA256 too.
+                data = (root_dir / path).read_bytes()
+                checksum_sha1 = hashlib.sha1(data).hexdigest()
+                checksum_sha256 = hashlib.sha256(data).hexdigest()
+
+                file_spdx_id = spdx_id(f"SPDXRef-FILE-{path}")
+                sbom_files.append({
+                    "SPDXID": file_spdx_id,
+                    "fileName": path,
+                    "checksums": [
+                        {"algorithm": "SHA1", "checksumValue": checksum_sha1},
+                        {"algorithm": "SHA256", "checksumValue": checksum_sha256}
+                    ]
+                })
+
+                # Tie each file back to its respective package.
+                sbom_relationships.append({
+                    "spdxElementId": package_spdx_id,
+                    "relatedSpdxElement": file_spdx_id,
+                    "relationshipType": "CONTAINS"
```

Comment:
Nit: trailing comma to minimise the diff when new items are added in the future. Ignore if desired :) 

```suggestion
                    "relationshipType": "CONTAINS",
```